src/dbnode/storage/series/buffer.go

```diff
@@ -89,90 +99,68 @@ type databaseBuffer interface {
 
 	Stats() bufferStats
 
-	// MinMax returns the minimum and maximum blockstarts for the buckets
-	// that are contained within the buffer. These ranges exclude buckets
-	// that have already been drained (as those buckets are no longer in use.)
-	MinMax() (time.Time, time.Time, error)
-
 	Tick() bufferTickResult
 
-	NeedsDrain() bool
-
-	DrainAndReset() drainAndResetResult
+	Bootstrap(bl block.DatabaseBlock)
 
-	Bootstrap(bl block.DatabaseBlock) error
-
-	Reset(opts Options)
+	Reset(blockRetriever QueryableBlockRetriever, opts Options)
 }
 
 type bufferStats struct {
-	openBlocks  int
 	wiredBlocks int
 }
 
-type drainAndResetResult struct {
-	mergedOutOfOrderBlocks int
-}
-
 type bufferTickResult struct {
 	mergedOutOfOrderBlocks int
 }
 
 type dbBuffer struct {
-	opts              Options
-	nowFn             clock.NowFn
-	drainFn           databaseBufferDrainFn
-	pastMostBucketIdx int
-	buckets           [bucketsLen]dbBufferBucket
-	blockSize         time.Duration
-	bufferPast        time.Duration
-	bufferFuture      time.Duration
+	opts           Options
+	nowFn          clock.NowFn
+	blockRetriever QueryableBlockRetriever
+
+	bucketsMap map[xtime.UnixNano]BufferBucketVersions
+	// Cache of buckets to avoid map lookup of above
+	bucketsCache       [bucketsCacheSize]BufferBucketVersions
```

Comment:
This should be an array of structs yeah? Otherwise they won't be alloc'd when this struct is alloc'd.